cc/gtp_player.cc

```diff
@@ -0,0 +1,323 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "cc/gtp_player.h"
+
+#include <iomanip>
+#include <iostream>
+#include <utility>
+
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_join.h"
+#include "absl/strings/str_split.h"
+#include "absl/time/clock.h"
+
+namespace minigo {
+
+namespace {
+
+bool CheckArgsExact(absl::string_view cmd, size_t expected_num_args,
+                    const std::vector<absl::string_view>& args) {
+  if (args.size() != expected_num_args) {
+    std::cerr << "ERROR: expected " << expected_num_args
+              << " args for GTP command \"" << cmd << "\", got " << args.size()
+              << " args:";
+    for (const auto& arg : args) {
+      std::cerr << " \"" << arg << "\"";
+    }
+    std::cerr << "\n";
+    return false;
+  }
+  return true;
+}
+
+bool CheckArgsRange(absl::string_view cmd, size_t expected_min_args,
+                    size_t expected_max_args,
+                    const std::vector<absl::string_view>& args) {
+  if (args.size() < expected_min_args || args.size() > expected_max_args) {
+    std::cerr << "ERROR: expected between " << expected_min_args << " and "
+              << expected_max_args << " args for GTP command \"" << cmd
+              << "\", got " << args.size() << " args:";
+    for (const auto& arg : args) {
+      std::cerr << " \"" << arg << "\"";
+    }
+    std::cerr << "\n";
+    return false;
+  }
+  return true;
+}
+
+}  // namespace
+
+GtpPlayer::GtpPlayer(std::unique_ptr<DualNet> network, const Options& options)
+    : MctsPlayer(std::move(network), options),
+      num_readouts_(options.num_readouts) {}
+
+bool GtpPlayer::HandleCmd(const std::string& line) {
+  std::vector<absl::string_view> args =
+      absl::StrSplit(line, absl::ByAnyChar(" \t\r\n"), absl::SkipWhitespace());
+  if (args.empty()) {
+    std::cout << "=" << std::endl;
```

Comment:
GTP spec calls for an "id" see 2.6
http://www.lysator.liu.se/~gunnar/gtp/gtp2-spec-draft2/gtp2-spec.html

Other implementations I've seen all add a print_gtp(int id, {var args, string}); method
see http://www.lysator.liu.se/~gunnar/gtp/gtp.c

Reply:
The id is optional according to 2.5 in the spec and I'm following the observed behaviour of our Python implementation.

Happy to add id support later along with more GTP features that I'm missing. This CL is just getting enough support in place to run Minigui (which does not expect IDs).